server/src/main/java/io/crate/collections/accountable/AccountableList.java

```diff
@@ -0,0 +1,189 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.collections.accountable;
+
+import static org.apache.lucene.util.RamUsageEstimator.NUM_BYTES_OBJECT_REF;
+
+import java.util.AbstractList;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.RandomAccess;
+import java.util.function.LongConsumer;
+
+import org.apache.lucene.util.RamUsageEstimator;
+
+/**
+ * List implementation accounting for shallow memory usage.
+ * Based on {@link ArrayList} with some simplifications.
+ * Doesn't have concurrent modification check.
+ * Supports boundary checks since some features (lag/lead) rely on IndexOutOfBoundsException.
+ * Supports only {@link #add(Object)} operation.
+ */
+public class AccountableList<T> extends AbstractList<T> {
+
+    private static final int SOFT_MAX_ARRAY_LENGTH = Integer.MAX_VALUE - 8;
+    private static final Object[] DEFAULTCAPACITY_EMPTY_ELEMENTDATA = {};
+    private static final int DEFAULT_CAPACITY = 10;
+
+    private final LongConsumer allocateBytes;
+    private Object[] elementData;
+    private int size;
+
+    public AccountableList(LongConsumer allocateBytes) {
+        this.allocateBytes = allocateBytes;
+        this.elementData = DEFAULTCAPACITY_EMPTY_ELEMENTDATA; // Empty now, will be accounted on the first growth.
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public T get(int index) {
+        Objects.checkIndex(index, size);
+        return (T) elementData[index];
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    /**
+     * Similar to {@link ArrayList#sort(Comparator)} but without modification count checks.
+     */
+    public void sort(Comparator<? super T> c) {
+        Arrays.sort((T[]) elementData, 0, size, c);
```

Comment:
As `Arrays.sort` state: `Temporary storage requirements vary from a small constant for nearly sorted input arrays to n/ 2 object references for randomly ordered input arrays.` shouldn't we account for n/2 object references here?

Reply:
Good catch, thanks. Added it to both list/sub-list sort implementations.